casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala

```diff
@@ -97,49 +92,31 @@ object DagOperations {
-    *
-    * TODO: Implement a GCA that doesn't require the genesis block (see https://rchain.atlassian.net/browse/RCHAIN-3002)
-    * TODO: Remove usage of BlockStore by just using BlockDagRepresentation (see https://rchain.atlassian.net/browse/RCHAIN-3003)
+    * We compute by finding the first block that is the "oldest" (has highest blocknum) common block.
     */
-  def greatestCommonAncestorF[F[_]: Monad: BlockStore](
-      b1: BlockMessage,
-      b2: BlockMessage,
-      genesis: BlockMessage,
+  def greatestCommonAncestorF[F[_]: Monad](
+      b1: BlockMetadata,
+      b2: BlockMetadata,
       dag: BlockDagRepresentation[F]
-  ): F[BlockMessage] =
+  ): F[BlockMetadata] = {
+    type BlockWithDepth = (Long, BlockMetadata)
+    implicit val blockNumDecreasing: Ordering[BlockWithDepth] =
+      (l: BlockWithDepth, r: BlockWithDepth) => r._1.compare(l._1)
+
+    def getParents(p: BlockMetadata): F[Set[BlockMetadata]] =
+      p.parents.map(v => dag.lookup(v)).sequence.map(_.toSet.flatten)
+
     if (b1 == b2) {
       b1.pure[F]
     } else {
-      def commonAncestorChild(
-          b: BlockMessage,
-          commonAncestors: Set[BlockMessage]
-      ): F[List[BlockMessage]] =
-        for {
-          childrenHashesOpt      <- dag.children(b.blockHash)
-          childrenHashes         = childrenHashesOpt.getOrElse(Set.empty[BlockHash])
-          children               <- childrenHashes.toList.traverse(ProtoUtil.unsafeGetBlock[F])
-          commonAncestorChildren = children.filter(commonAncestors)
-        } yield commonAncestorChildren
-
-      for {
-        b1Ancestors     <- bfTraverseF[F, BlockMessage](List(b1))(ProtoUtil.unsafeGetParents[F]).toSet
-        b2Ancestors     <- bfTraverseF[F, BlockMessage](List(b2))(ProtoUtil.unsafeGetParents[F]).toSet
-        commonAncestors = b1Ancestors.intersect(b2Ancestors)
-        gca <- bfTraverseF[F, BlockMessage](List(genesis))(commonAncestorChild(_, commonAncestors))
-                .findF(
-                  b =>
-                    for {
-                      childrenOpt <- dag.children(b.blockHash)
-                      children    = childrenOpt.getOrElse(Set.empty[BlockHash]).toList
-                      result <- children.existsM(
-                                 hash =>
-                                   for {
-                                     c <- ProtoUtil.unsafeGetBlock[F](hash)
-                                   } yield b1Ancestors(c) ^ b2Ancestors(c)
-                               )
-                    } yield result
-                )
-      } yield gca.get
+      val start: SortedSet[BlockWithDepth] = SortedSet
+        .empty[BlockWithDepth] + ((b1.blockNum, b1)) + ((b2.blockNum, b2))
+      Monad[F]
+        .iterateWhileM(start)(s => {
+          val (head, tail) = (s.head._2, s.tail)
+          getParents(head).map(nb => tail ++ nb.map(v => (v.blockNum, v)))
+        })(_.size != 1)
+        .map(v => v.head._2)
```

Comment:
I haven't really tried hard yet but I would appreciate if you could make this more readable somehow.